### PR TITLE
Add return-types with help from DebugClassLoader in the CI

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #33228
| License       | MIT
| Doc PR        | -

I've spent a great deal of time on this PR, experimenting with adding return types to the codebase.

TL;DR: my conclusion is that we cannot make it for 5.0.

There are two reasons for this:
1. The burden this will put on the community is immense, especially when considering that third party libs must also be updated for any apps to work at all on a return-typed 5.0. Symfony must add them last, not first.
2. We need return type covariance, yet this won't be available before PHP 7.4, while 5.0 supports 7.2.

What's attached?
- ~a draft patching logic in `DebugClassLoader` to add return-type where it discovers this should be done~
- return types added automatically thanks to #33283
- ~manual fixes for situations not handled (yet, if possible at all) by that logic in `DebugClassLoader`~ #33332

What's achieved? Tests are green \o/

At this stage, I think we have to acknowledge we won't add return-types in 5.0 but prepare a serious plan to add them in 6.0.

This plan could be:
- [x] make DebugClassLoader able to automate adding return types.
- [x] in 4.4: add all possible return types that don't break BC, e.g. in `Tests` and in generated code
- [x] spot and fix places where annotations aren't accurate, add more annotations where possible.
- [x] ensure `DebugClassLoader` triggers the best possible deprecations that encourage ppl to add return-types in their libs/apps. This means we could decide to disable the current ones (see #33235) and to re-enable them in 5.1. This will also give us the time to fine-tune the tooling (item 1. on this list)

Ideally, we could reach a point where we could test branch 4.4 *with* return-types: we'd use the tooling to add them automatically in the CI job, then we'd run tests and they should be green. Let's do this?

Help Wanted, here is how:

*With PHP 7.4*, run `php .github/patch-types.php`. This will add return types everywhere possible.

Then run tests, e.g. `./phpunit src/Symfony/Component/HttpFoundation --exclude-group legacy,issue-32995`

Here are the components that fail with return types added, please help me check them all with a PR on [my fork](https://github.com/nicolas-grekas/symfony/tree/eh-return-types):
 - [x] src/Symfony/Bridge/Doctrine
 - [x] src/Symfony/Bridge/Monolog
 - [x] src/Symfony/Bridge/PhpUnit
 - [x] src/Symfony/Bridge/ProxyManager
 - [x] src/Symfony/Bridge/Twig
 - [x] src/Symfony/Bundle/DebugBundle
 - [x] src/Symfony/Bundle/FrameworkBundle
 - [x] src/Symfony/Bundle/SecurityBundle
 - [x] src/Symfony/Bundle/TwigBundle
 - [x] src/Symfony/Bundle/WebProfilerBundle
 - [x] src/Symfony/Bundle/WebServerBundle
 - [x] src/Symfony/Component/Asset
 - [x] src/Symfony/Component/BrowserKit
 - [x] src/Symfony/Component/Cache
 - [x] https://github.com/nicolas-grekas/symfony/pull/28 src/Symfony/Component/Config
 - [x] src/Symfony/Component/Console
 - [x] src/Symfony/Component/CssSelector
 - [x] src/Symfony/Component/Debug
 - [x] https://github.com/nicolas-grekas/symfony/pull/28 src/Symfony/Component/DependencyInjection
 - [x] src/Symfony/Component/DomCrawler
 - [x] src/Symfony/Component/Dotenv
 - [x] src/Symfony/Component/ErrorHandler
 - [x] src/Symfony/Component/ErrorRenderer
 - [x] https://github.com/nicolas-grekas/symfony/pull/24 src/Symfony/Component/EventDispatcher
 - [x] src/Symfony/Component/ExpressionLanguage
 - [x] src/Symfony/Component/Filesystem
 - [x] src/Symfony/Component/Finder
 - [x] src/Symfony/Component/Form
 - [x] src/Symfony/Component/HttpClient
 - [x] src/Symfony/Component/HttpFoundation
 - [x] src/Symfony/Component/HttpKernel
 - [x] src/Symfony/Component/Inflector
 - [x] src/Symfony/Component/Intl
 - [x] src/Symfony/Component/Ldap
 - [x] src/Symfony/Component/Lock
 - [x] src/Symfony/Component/Mailer
 - [x] src/Symfony/Component/Messenger
 - [x] src/Symfony/Component/Mime
 - [x] src/Symfony/Component/OptionsResolver
 - [x] src/Symfony/Component/Process
 - [x] src/Symfony/Component/PropertyAccess
 - [x] src/Symfony/Component/PropertyInfo
 - [x] https://github.com/nicolas-grekas/symfony/pull/25 src/Symfony/Component/Routing
 - [x] https://github.com/nicolas-grekas/symfony/pull/26 src/Symfony/Component/Security
 - [x] src/Symfony/Component/Security/Core
 - [x] src/Symfony/Component/Security/Guard
 - [x] src/Symfony/Component/Security/Http
 - [x] https://github.com/nicolas-grekas/symfony/pull/29 src/Symfony/Component/Serializer
 - [x] src/Symfony/Component/Security/Csrf
 - [x] src/Symfony/Component/Stopwatch
 - [x] src/Symfony/Component/Templating
 - [x] https://github.com/nicolas-grekas/symfony/pull/27 src/Symfony/Component/Translation
 - [x] src/Symfony/Component/Validator
 - [x] src/Symfony/Component/VarDumper
 - [x] src/Symfony/Component/VarExporter
 - [x] src/Symfony/Component/WebLink
 - [x] src/Symfony/Component/Workflow
 - [x] src/Symfony/Component/Yaml
 - [x] src/Symfony/Contracts

### CODE REVIEW

**Comment by fabpot** [comment]
Sounds like a solid plan to me. We are not in a hurry. We need to carefully build a plan and targeting 6.0 is probably the thing to do.

**Comment by derrabus** [comment]
First of all, thank you very much for all the effort you have put into this!

> TL;DR: my conclusion is that we cannot make it for 5.0.

That's a sad and disappointing conclusion. 😞 But I refuse to give up at this point – especially after the wall of text that I posted last night as #33228. 😉 We might not get fully typed, but I'm still convinced we can get further than we are at the moment.

We already have added a few return types (mainly because we were somewhat forced to: #27905) and probably will be required to add a few more (#33039).

> 1. The burden this will put on the community is immense,

Yes, but that'll still be the case two years from now. The main difference I see is that we won't have a maintained Symfony version anymore that still supports php 5. I admit, that might help a little. 🤔 

> especially when considering that third party libs must also be updated for any apps to work at all on a return-typed 5.0.

Why's that? Having types in Symfony does not require the whole world to release typed versions of their libraries as well.

Third-party libraries only need to be updated if they implement/extend Symfony interfaces/classes that received types with 5.0. Perhaps we can find some problematic hot spots and exclude them when adding return types, such as (list might be incomplete):

* `Symfony\Component\Config\Definition\ConfigurationInterface`
* `Symfony\Component\DependencyInjection\Extension\ExtensionInterface`
* `Symfony\Component\EventDispatcher\EventSubscriberInterface`
* `Symfony\Component\HttpKernel\Bundle\BundleInterface`

That could ease the migration for many third-party bundles.

> 2. We need return type covariance, yet this won't be available before PHP 7.4, while 5.0 supports 7.2.

Can you elaborate on this point? I mean, having full covariance is certainly nice to have, but wouldn't call it a requirement. Apart from fluent interfaces (for which we have found a reasonable compromise), where would we actually depend on covariance? Maybe we can exclude those classes when adding return types or work with a combination of a more generic return type declaration and a more specific `@return` annotation here? We should even be able to teach `DebugClassLoader` to detect such cases.

I mean, there might always be features in newer php releases that might come handy. Who knows, maybe in two years from now, generics have been added to php 8.1 and then we say: "Hey, we can't add return types because we desperately need generics to go fully typed!"

> * in 4.4: add all possible return types that don't break BC, e.g. in `Tests` and in generated code

For 5.0: At least add return types to everything that is flagged as `@internal`? 🥺

> * spot and fix places where annotations aren't accurate, add more annotations where possible.

I expect to find most of these cases when actually adding return types. 🤔 

> Ideally, we could reach a point where we could test branch 4.4 _with_ return-types: we'd use the tooling to add them automatically in the CI job, then we'd run tests and they should be green. Let's do this?

Yes! Even if we will never add types, we would end up with properly configured mocks and a codebase without inconsitent return points. Wouldn't be too bad after all. 😎 

In preparation of the issue I wrote yesterday, I've done some experiments on my own. I followed a more manual approach, using PhpStorm inspections and refactorings. One result that I already have is a typed version of the EventDispatcher. I will post a PR later – even if we don't merge it, it might still be a good reference. Also, we might want to cherry-pick the changes I've made to the tests. 😃

**Comment by nicolas-grekas** [comment]
> That's a sad and disappointing conclusion.

The issue is complex enough, let's keep a cool head. I don't think it's not sad when it's a well-crafted engineering decision, and when there's a plan forward to bring return-types in a less brutal way.

> We already have added a few return types (mainly because we were somewhat forced to: #27905) and probably will be required to add a few more (#33039).

True, that's the cost of having third-party dependencies. Yet the scope is very limited, it's not the same story than doing it for the whole codebase.

>> The burden this will put on the community is immense,
> Yes, but that'll still be the case two years from now.

Absolutely not, you missed the last item on my plan above: we would throw deprecation notices starting from 5.1 (the same as `DebugClassLoader` does right now), so that ppl would be encouraged to add return types well ahead of v6.

> Having types in Symfony does not require the whole world to release typed versions of their libraries as well.
> Third-party libraries only need to be updated if they implement/extend Symfony interfaces/classes
 
That's enough a blocker. ALL lines changed in this PR are individual BC breaks. That's way too abrupt. This can make v5 fail totally.

> having full covariance is certainly nice to have, but wouldn't call it a requirement.

That's what I was reminded when working on this PR: e.g. `FrameworkBundle`'s `Client::doRequest()` returns `Response`, while the parent returns `object`. That's totaly legal. Form has similar needs for covariance, DoctrineBridge too. DebugClassLoader can detect them, but cannot change this: if ppl add the parent's return type as they would be forced to do if we add them, it will be impossible to add those return types ever.

> For 5.0: At least add return types to everything that is flagged as @internal? 🥺

Yes, or 4.4.

> Even if we will never add types, we would end up with properly configured mocks and a codebase without inconsistent return points. Wouldn't be too bad after all.

:sunglasses:

**Comment by Tobion** [comment]
>> For 5.0: At least add return types to everything that is flagged as @Internal? 🥺

> Yes, or 4.4.

We could also add return types to final classes, ref. #33152

**Comment by derrabus** [comment]
> > That's a sad and disappointing conclusion.
> 
> The issue is complex enough, let's keep a cool head. I don't think it's sad when it's a well-crafted engineering decision, and when there's a plan forward to bring return-types in a less brutal way.

Sure, I meant no offense. After all the effort that has gone into this already, of course I am disappointed if the conclusion is that we won't proceed for now. That doesn't mean that you've made a bad decision. Your plan sounds reasonable and we really might be in a better position two years from now.

Yet, you make it sound like we have to decide between two extreme options: Leaving everything as it is or going 100%. I was trying to find a middle ground. Adding types where the impact is low. Of course, estimating the impact of each change is hard and nothing we could do fully automated.

But this is ultimately not my decision. If the plan is to make adding return types a smooth two-year project, I'm still in and would offer my help with executing the plan, if I may.

Also, the more I think about your suggestion of a CI run with an automatically patched typed version of the current codebase, the more I like it. That way, we might already get some benefits of adding return types without actually adding them. Can you that that up? I would offer to help fixing those tests.

> ALL lines changed in this PR are individual BC breaks. That's way too abrupt. This can make v5 fail totally, as far as adoption is concerned.

Yes, you are right. I might have underestimated that danger.

**Comment by derrabus** [comment]
> One result that I already have is a typed version of the EventDispatcher. I will post a PR later – even if we don't merge it, it might still be a good reference.

As promised: #33248

**Comment by nicolas-grekas** [comment]
This PR now paves the way execute the plan: attached are some scripts to add return types in the CI, just before running the tests. This fails for now, but we should be able to fix them all. I already fixed some issues actually, and would happily accept PRs [on my fork](https://github.com/nicolas-grekas/symfony/tree/eh-return-types) to move this forward! Can't wait your PRs :)

**Comment by nicolas-grekas** [comment]
Help Wanted, here is how:

*With PHP 7.4*, run `php .github/patch-types.php`. This will add return types everywhere possible.

Then run tests, e.g. `./phpunit src/Symfony/Component/HttpFoundation --exclude-group legacy`

Here are the components that fail with return types added, please help me check them all with a PR on [my fork](https://github.com/nicolas-grekas/symfony/tree/eh-return-types):
 - [ ] src/Symfony/Bridge/Doctrine
 - [ ] src/Symfony/Bridge/ProxyManager
 - [ ] src/Symfony/Bridge/Twig
 - [ ] src/Symfony/Bundle/FrameworkBundle
 - [ ] src/Symfony/Bundle/SecurityBundle
 - [ ] src/Symfony/Bundle/TwigBundle
 - [ ] src/Symfony/Bundle/WebProfilerBundle
 - [ ] src/Symfony/Component/BrowserKit
 - [ ] src/Symfony/Component/Cache
 - [ ] src/Symfony/Component/Config
 - [ ] src/Symfony/Component/Console
 - [ ] src/Symfony/Component/DependencyInjection
 - [ ] src/Symfony/Component/DomCrawler
 - [ ] src/Symfony/Component/ErrorHandler
 - [ ] src/Symfony/Component/EventDispatcher
 - [ ] src/Symfony/Component/Form
 - [ ] src/Symfony/Component/HttpKernel
 - [ ] src/Symfony/Component/Intl
 - [ ] src/Symfony/Component/Messenger
 - [ ] src/Symfony/Component/PropertyAccess
 - [ ] src/Symfony/Component/PropertyInfo
 - [ ] src/Symfony/Component/Routing
 - [ ] src/Symfony/Component/Security
 - [ ] src/Symfony/Component/Security/Core
 - [ ] src/Symfony/Component/Security/Guard
 - [ ] src/Symfony/Component/Security/Http
 - [ ] src/Symfony/Component/Serializer
 - [ ] src/Symfony/Component/Templating
 - [ ] src/Symfony/Component/Translation
 - [ ] src/Symfony/Component/Validator
 - [ ] src/Symfony/Component/Workflow
 - [x] src/Symfony/Bridge/Monolog
 - [x] src/Symfony/Bridge/PhpUnit
 - [x] src/Symfony/Bundle/DebugBundle
 - [x] src/Symfony/Bundle/WebServerBundle
 - [x] src/Symfony/Component/Asset
 - [x] src/Symfony/Component/CssSelector
 - [x] src/Symfony/Component/Debug
 - [x] src/Symfony/Component/Dotenv
 - [x] src/Symfony/Component/ErrorRenderer
 - [x] src/Symfony/Component/ExpressionLanguage
 - [x] src/Symfony/Component/Filesystem
 - [x] src/Symfony/Component/Finder
 - [x] src/Symfony/Component/HttpClient
 - [x] src/Symfony/Component/HttpFoundation
 - [x] src/Symfony/Component/Inflector
 - [x] src/Symfony/Component/Ldap
 - [x] src/Symfony/Component/Lock
 - [x] src/Symfony/Component/Mailer
 - [x] src/Symfony/Component/Mime
 - [x] src/Symfony/Component/OptionsResolver
 - [x] src/Symfony/Component/Process
 - [x] src/Symfony/Component/Security/Csrf
 - [x] src/Symfony/Component/Stopwatch
 - [x] src/Symfony/Component/VarDumper
 - [x] src/Symfony/Component/VarExporter
 - [x] src/Symfony/Component/WebLink
 - [x] src/Symfony/Component/Yaml
 - [x] src/Symfony/Contracts

File: src/Symfony/Component/HttpKernel/Profiler/Profile.php [hunk 1]
```diff
@@ -134,13 +134,16 @@ public function setMethod($method)
...
-     * @return string The URL
+     * @return string|null The URL
      */
     public function getUrl()
     {
         return $this->url;
     }
 
+    /**
+     * @param $url string
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/Profiler/Profile.php]: `string $url`

File: src/Symfony/Component/Ldap/Security/LdapUserProvider.php [hunk 2]
```diff
@@ -144,7 +144,7 @@ public function supportsClass($class)
     /**
      * Loads a user from an LDAP entry.
      *
-     * @return LdapUser
+     * @return LdapUser|UserInterface
```
Comment (ro0NL) [inline src/Symfony/Component/Ldap/Security/LdapUserProvider.php]: why both? The real return type would still be one-of

Comment (nicolas-grekas) [reply src/Symfony/Component/Ldap/Security/LdapUserProvider.php]: > Declaration of Symfony\Component\Security\Core\User\LdapUserProvider::loadUser($username, Symfony\Component\Ldap\Entry $entry): Symfony\Component\Security\Core\User\User must be compatible with Symfony\Component\Ldap\Security\LdapUserProvider::loadUser($username, Symfony\Component\Ldap\Entry $entry): Symfony\Component\Ldap\Security\LdapUser in /var/www/html/symfony/src/Symfony/Component/Security/Core/User/LdapUserProvider.php on line 55

File: src/Symfony/Component/HttpKernel/Profiler/Profile.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/Profiler/Profile.php]: fixed thx

File: src/Symfony/Component/Ldap/Security/LdapUserProvider.php (hunk 2 above)
Comment (ro0NL) [reply src/Symfony/Component/Ldap/Security/LdapUserProvider.php]: i see, missed that. But this new ldap provider we aim for `loadUser(): UserInterface` in 5.0 isnt it? shouldnt we hint as such already?

Comment (nicolas-grekas) [reply src/Symfony/Component/Ldap/Security/LdapUserProvider.php]: right, updated

File: src/Symfony/Component/Security/Guard/AuthenticatorInterface.php [hunk 3]
```diff
@@ -90,7 +90,7 @@ public function getUser($credentials, UserProviderInterface $userProvider);
      *
      * @param mixed $credentials
      *
-     * @return bool
+     * @return mixed true if the credentials are valid, anything else otherwise.
```
Comment (derrabus) [inline src/Symfony/Component/Security/Guard/AuthenticatorInterface.php]: Now that #33308 has been merged, we should revert this change.

Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Guard/AuthenticatorInterface.php]: reverted

**Comment by nicolas-grekas** [comment]
PR now green on Travis, help wanted for fixing failures on appveyor.
